namer/consul/src/main/scala/io/buoyant/namer/consul/LookupCache.scala

```diff
@@ -33,59 +39,72 @@ private[consul] class LookupCache(
+            val observationClosable = addr.changes.respond {
+              case Addr.Neg => observationState.update(Activity.Ok(NameTree.Neg))
+              case Addr.Pending => observationState.update(Activity.Pending)
+              case Addr.Failed(why) => observationState.update(Activity.Failed(why))
+              case Addr.Bound(_, _) => observationState.update(Activity.Ok(NameTree.Leaf(Name.Bound(addr, id, residual))))
+            }
+            Future.value(observationClosable)
+          case Throw(cause) =>
+            // We probably failed to fetch agent config. This is critical.
+            // TODO: if this has happened only restart can restore consul namer to working state. Throw exception instead?
+            observationState.update(Activity.Failed(cause))
+            Future.value(Closable.nop)
         }
 
-        val observation = Var.async[Activity.State[NameTree[Name.Bound]]](Activity.Pending) { observationState =>
-          val closableFuture = addrFuture.transform {
-            case Return(addr) =>
-              val observationClosable = addr.changes.respond {
-                case Addr.Neg => observationState.update(Activity.Ok(NameTree.Neg))
-                case Addr.Pending => observationState.update(Activity.Pending)
-                case Addr.Failed(why) => observationState.update(Activity.Failed(why))
-                case Addr.Bound(_, _) => observationState.update(Activity.Ok(NameTree.Leaf(Name.Bound(addr, id, residual))))
-              }
-              Future.value(observationClosable)
-            case Throw(cause) =>
-              // We probably failed to fetch agent config. This is critical.
-              // TODO: if this has happened only restart can restore namerd to working state. Throw exception instead?
-              observationState.update(Activity.Failed(cause))
-              Future.value(Closable.nop)
-          }
-
-          Closable.make { deadline =>
-            closableFuture.flatMap(_.close(deadline))
-          }
+        Closable.make { deadline =>
+          closableFuture.flatMap(_.close(deadline))
         }
+      }
+
+      cachedCounter.incr()
+      InstrumentedBind(instrumentedObsv, pollState)
+    }
 
-        cachedCounter.incr()
-        new Activity(observation)
-    })
-
-  def apply(
-    dc: String,
-    key: SvcKey,
-    id: Path,
-    residual: Path
-  ): Activity[NameTree[Name]] = {
-    log.debug("consul lookup: %s %s", dc, id.show)
-    lookupCounter.incr()
-    cachedLookup(dc, key, id, residual)
+  def apply(servicePath: Path): Option[Activity[NameTree[Name]]] = {
+    val bound =
+      Option(lookupStatus.get(servicePath)) orElse
+        pathParser(servicePath).map {
+          case (dc, key, id, residual) =>
+            log.debug("consul lookup: %s %s", dc, id.show)
+            lookupCounter.incr()
+            val binding = lookup(dc, key, id, residual)
+            lookupStatusMu.synchronized { lookupStatus.put(servicePath, binding) }
```

Comment:
I think we need to expand the scope of this synchronized block to include the creation of the Activity (ie the call to lookup) and also add a second check to see if the entry is in the cache inside the synchronized block.

This is to avoid the race condition where two different Activities could be returned for the same path.